### ISSUE TITLE
Validate only one field at a time 

### ISSUE THREAD

**Issue by orenklein** [jaredpalmer/formik#512]
## Bug, Feature, or Question?
Feature/Question

## Current Behavior
Currently, validation is running asynchronously on the whole schema (assuming you use ValidationSchema of yup). It is reasonable to run the validation on submit, however, It is extremely cumbersome to run it on every field change/blur. 

Especially if you have some backend validation (using yup.test). 
Assuming you have a couple of fields, field A, and field B - the later must go through backend validation. Every time you change field A, the validation runs both on A and B which will cause unnecessary backend calls.

I also tried using the new alpha version's handleChange('fieldName') but I still experience the same behavior.

## Suggested Solutions

1. Using yup schema, using yup.reach seems reasonable, even though I'm not sure how is its performance (https://github.com/jaredpalmer/formik/issues/35)

2. Formik validate function - pass the event's field name. It will allow the developer to validate only one field at a time.

## Environment
- Formik Version: 0.11.11 /1.0.0-alpha.2
- React Version: 16.2
- TypeScript Version: 2.7.2
- CodeSandbox Link: 
- OS: macOS High Sierra
- Node Version: 8.9.4
- Package Manager and Version:

**Comment by jaredpalmer** [comment]
We could prototype this with a special version of Field. The tougher bit is how to do it without Yup. 

Open to PRs. Also FYI/reference, unlike Field, FastField tries to run validation synchronously before running it async. Perhaps that technique should be moved as the default

**Comment by jquense** [comment]
For the general case, it seems like just passing `fieldName` to `validate` is "good enough" to let someone handle that if they want. You can implement validationSchema in terms of that api for single field validations. It's not particularly hard to do with yup, `react-formal` implements it like this: https://github.com/jquense/react-formal/blob/master/src/Form.js#L368-L383

**Comment by jaredpalmer** [comment]
Interesting. Will check this out.

**Comment by JonathanStoye** [comment]
Any updates on this? Otherwise I would give @jquense idea of passing the `fieldName` to `validate` a shot if you don't mind.

**Comment by sshmyg** [comment]
Any solutions?

**Comment by sshmyg** [comment]
’test’ method of yup schema invokes even changes some other field. It's very inconvenient.

**Comment by jaredpalmer** [comment]
You can pass a validate function to Field

--
Jared
________________________________

**Comment by sshmyg** [comment]
Oh, you mean I don't need `.test` in yup schema and just validate in every require field separately?

**Comment by sshmyg** [comment]
Validate all schema onChange, it's very bad solution.

**Comment by jaredpalmer** [comment]
See 1.0.0-beta.3 Field validate.

**Comment by sshmyg** [comment]
I think it's not an option use some Field component.
Question was about `yup.test` validation.
You should warn in readme than about using yup.
`test` and 'transform' doesn't work as expected.

**Comment by suberg** [comment]
@jaredpalmer if I pass `validate` function to my Field components, it runs all these functions when I edit only one field. Why? It's not obvious. I don't want to use Yup. Is there a possibility to disable this behaviour and run one validate function?

**Comment by idesignpixels** [comment]
Would also like a solution to this, my scenario we query a pay as you go api for validation and this would call it over and over with the same value unless we intervene.

**Comment by jaredpalmer** [comment]
@suberg @idesignpixels let me make sure I am understanding the suggested behavior:

- Formik `validationSchema` 
  - Runs only validation for each field's key onChange/onBlur or equivalent imperative methods. 
  - Runs all as full schema prior to submit
- Formik `validate` 
  - Stays the same as this is offloaded to userland
- Field `validate`
   - Stays same, single `value` is first argument

**Comment by ivanquirino** [comment]
@jaredpalmer let's say we have `username` and `email` fields. We set our validation schema to 
```
yup.object().shape({
  email: yup.string().required().email()
}}
```
Then we pass a custom validation function as validate prop to the username Field that validates if our username is unique. What is happening is that when I type at the email field, the validate function provided to the username field is also running.

**Comment by asgeirfri** [comment]
I think I have an ok solution for people using <Formik /> and normal input fields

`
import { reach } from 'yup'

// Custom function to validate single field in validationSchema
  validateSingle = (name, formProps) => {
    reach(validationSchema, name).validate( formProps.values[name] )
      .then(() => {
        formProps.setFieldError(name, '')
      }).catch(error => {
        formProps.setFieldError(name, error.message)
      })
  }`

Calling this onBlur should only update the single error field and run the single validation
Just pass in the name of the field and the formProps

[reach](https://www.npmjs.com/package/yup#yupreachschema-schema-path-string-value-object-context-object-schema)
[setFieldError](https://jaredpalmer.com/formik/docs/api/formik#setfielderror-field-string-errormsg-string-void) 

This may need some editing for nested validation tho :)

**Comment by newsroomdev** [comment]
Question @jaredpalmer, are you saying current the current expected behavior for `validate` on Field components should only run when that specific field is touched? I am using only Field-level validate props with Formik v1.5.1, but I'm also still seeing all validations re-running on change and blur.

**Comment by Andreyco** [comment]
I think it makes sense to
- run validation on single field if `validate` is specified on `Field` component
- run validation on all field if `validate` is specified on nearest `Formik` component
- run validations on all fields on submit, regardless where `validate` is specified, be it `Field` or `Formik` component

@jaredpalmer  what do you think?

**Comment by viljark** [comment]
Is there built in way to run only single field validation on change/blur? 

@jaredpalmer has stated that 
"However Formik supports Field-level validation now, which does only run one-at-a-time."

but I'm unable to find any examples and for me all the <Field>s run their validation functions on any field change/blur.

**Comment by acasaccia** [comment]
I had the same problem as others here, Formik set up with Yup and `validationSchema`, and a `test` async validation that was repeated for each keystroke, on other fields too.

I worked this around by invoking my remote validation `onBlur`:

```
onBlur={(e) => {
	handleBlur(e);
	$.get('/api/validations/companyName', { value: e.target.value }).then(response => {
		companyNameUnique = response.success;
		validateForm();
	});
}}
```

I stored the result of validation in a module variable (`companyNameUnique`) and then forced immediately a full form validation.

This way I can have a sync test in Yup that's simply using the cached result:

```
companyName: yup.string()
	.test('companyNameUnique', l.companyNameUnique, () => companyNameUnique),
```

**Comment by pedroabreu** [comment]
I'm running into the same issue, where all the fields that have been touched are validated onChange and onBlur on any field.

I'm using the `validate` on the `Field` with `name` prop set.

```
return (
        <Field
            name={name}
            validate={validateFn}
            render={({ field, form }) => (
                <CustomField
                    className={className}
                    childNode={children}
                    displayError={displayError}
                    label={label}
                    labelClassName={labelClassName}
                    meta={{ error: form.errors[field.name], touched: form.touched[field.name] }}
                    required={required}
                />
            )}
        />
    )
```
validateFn is an async function

From what I could check in the formik code, the validators run for the entire formik state (which is set on touched, on blur and on change). Instead of running it on the entire state, would it make sense to just run it for the field only instead ?

https://github.com/jaredpalmer/formik/blob/master/src/Formik.tsx#L129 

Besides on submission, what was the use case to always validate all the fields ? 🤔

**Comment by pedroabreu** [comment]
I've digged a bit into the docs and code and manage to figure one workaround for single field validation. It seems to be pretty hardcoded to always validate all fields set in state (and touched).

Start by disabling form level validation onChange and onBlur

```
<Formik
    ....
    validateOnChange={false}
    validateOnBlur={false}
>
```
This will avoid all field validations but will still set the values in the state (plus touched state)

There's an exposed prop called `validateField` that performs that single field validation once

```
const onInputBlur = (event, field) => {
    handleBlur(event)
    validateField(field)
}
```

validateField will handle setting the error in the formik state. So we do the formik `handleBlur` to set value and touched in state and then call the field validation with the field name.

This way also keeps the full validation on submit (by default).

@jaredpalmer do you think it's a reasonable workaround ? Haven't tested it fully but so far it's been avoiding API validation calls on certain fields (my goal). Maybe I'm missing some case that I haven't thought of.

**Comment by fenos** [comment]
@pedroabreu it could be a workaround; But just for people that use `field level` validation.

If you are using `validationSchema` and `yup` the `validateField` function is not an option. 

This behaviour of validating fields one by one is very much needed and it should have been the default from the beginning. 

At least we should start giving an option for opting out specific fields to be validated in case other fields are changed.

Then trigger validation for that specific field only when interacting with it.

Does anyone have a good workaround for this?!

**Comment by KosGrillis** [comment]
@pedroabreu 

You state
> There's an exposed prop called `validateField` that performs that single field validation once

What do you mean by this? Where is that prop exposed?

**Comment by pedroabreu** [comment]
In the `Formik` component at least. Here it is as part of an example for single field validationhttps://github.com/jaredpalmer/formik/blob/6899dc5cbd5cc9a818eae5cebbaca9b859483c35/examples/FieldLevelValidation.js

**Comment by jjalonso** [comment]
@asgeirfri 

Based on some experiences with Yup last week, your method will not work when you have Yup schemas that contains when(), and it depends of others, you need to send contexts etc... a problem.

**Comment by TidyIQ** [comment]
Any updates? I see this issue has gone stale...

**Comment by erassynathingo** [comment]
i've have a workaround incase anyone is still interested in this.

`
 <Form.Field error={errors.valueName && touched.valueName !== undefined ? true: false}>
                    <label>valueName</label>
                    <Input
                        value={values.valueName}
                        name="valueName"
                        error={errors.valueName && touched.valueName !== undefined ? true: false}
                        onChange={handleChange}
                        onBlur={handleBlur}
                     />
                     {errors.valueName && touched.valueName && <ErrorLabel 
                      children{errors.valueName}/>}
</Form.Field>
`

**Comment by cigzigwon** [comment]
@pedroabreu I agree but it looks like that would produce a new issue when dealing with a race condition, in my case it seems to be referencing prevState when using onChange handler instead of onBlur

**Comment by cigzigwon** [comment]
@jaredpalmer and @pedroabreu the better workaround would be using custom onChange handler with setFieldValue('borrower_additional', e.target.value, true) which would trigger validation correctly but it looks like using true there does not matter if validateOnChange is false. This seems to be the biggest problem w/Formik because I like others don't use a validation schema and don't need the whole visible form to validate on every single change

**Comment by cigzigwon** [comment]
@pedroabreu avoid the race condi w/
```
onChange={async e => {
		        		await handleChange(e)
		        		validateField('borrower_additional')
		        	}}
```

**Comment by pedroabreu** [comment]
@cigzigwon I've just spend several hours trying to figure out why was the value validated first and only after set in the state. I already had a codesandbox example to open an issue 😅 

It's weird that I didn't ran into that when on `onBlur` but only now on the `onChange`. I'll update the other post I made

**Comment by silvioBi** [comment]
Any update on this one?

**Comment by ops1ops** [comment]
Hi guys, i've recently started using formik and now my problem is that when 1 field changes validation goes on whole form fields so whole errors appear at props and my main questions is there any solution for this?

**Comment by sshmyg** [comment]
@ops1ops Just use some other lib.

**Comment by johnrom** [comment]
@ops1ops errors should not appear on the form because the error message component provided by Formik checks to see if a field has been touched (`formikProps.touched.myFieldName`). If you're using your own message display, you should also check the `touched` property to only display errors when it is true. Here's an example:

```
<Formik 
    initialValues={getInitialValues()} 
    validationSchema={getValidationSchema()}
>
    {formikProps => (
        <Form>
            <div>
                <Field name="myField" required />
                {
                    formikProps.touched.myField && 
                    formikProps.errors.myField && 
                    <p>{formikProps.errors.myField}</p>
                }
                // or
                <ErrorMessage name="myField" />
            </div>
        </Form>
    )}
</Formik>
```

This is documented here: https://jaredpalmer.com/formik/docs/api/errormessage

**Comment by ops1ops** [comment]
@johnrom thanks for your answer.
This is good example, but not really what i need. Currently i have to validate fields onChange, so if i use touched prop i cant really do it, because onChange validation will work only when field was touched. The solution i found is to make your own onChange handler using setFieldValue with setFieldTouched. And now i'm stuck with problem that formik updates all errors when you change field with setFieldValue.

**Comment by johnrom** [comment]
@ops1ops a field is marked as touched after its first change if you use the API provided by formik. if you use `setFieldValue`, this is not the case. But it is if you use `handleChange`. Formik will update all errors when you change a field with setFieldValue, but they will not display on a field until that field is both touched _and_ invalid.  If your issue is that there is a performance concern from changing all of the errors, I can understand that. If, however, your issue is that all of the errors will display at once, they should not if it is hooked up normally.

**Comment by ops1ops** [comment]
@johnrom 
> a field is marked as touched after its first change if you use the API provided by formik. 

are you sure about this? For example:
 
```
<Form.Item
    name="name"
    validateStatus={ touched.name && errors.name ? 'error' : 'success' }
    hasFeedback={ touched.name || values.name }
    help={ touched.name && errors.name }
>
    <input
        placeholder="name"
        name="name"
        // onChange={ (a) => {
        //     setFieldValue([a.target.name], a.target.value);
        //     setFieldTouched(a.target.name)
        // }}
        onChange={handleChange}
        onBlur={ handleBlur }
    />
</Form.Item>
```
Because even with this default input when you focus it and start changing, errors still dont appear because touched.name is not true. Maybe i dont know something, i've already said that i'm new with formik

**Comment by johnrom** [comment]
@ops1ops sorry I misspoke, yes it doesn't set touched until after the field is blurred. This is the default behavior in many forms so that the user is not presented with errors before they've had a chance to finish typing.

**Comment by ops1ops** [comment]
@johnrom so i was just interested is there anything in formik that provides onChange validation without being blurred. I managed this workaround         
```
        // onChange={ (a) => {
        //     setFieldValue([a.target.name], a.target.value);
        //     setFieldTouched(a.target.name)
        // }}
```
but i now i have another problems because of it.

**Comment by johnrom** [comment]
@ops1ops that's a slightly different topic than this issue, feel free to open a separate issue or join the discord to ask your question. if you open an issue please use the `Bug Report` template as it has information regarding creating a sandbox and filling out your Formik version information.

**Comment by ops1ops** [comment]
@johnrom ok, thanks for all

**Comment by clu7ch3r** [comment]
I've found a workaround. You should only display an error message when the text input field changes. Luckily, formik has a method called `getFieldMeta` for that and when you pass the field name to it, it returns an object. The only thing that changes in that method is the `value` prop which you would do a check on.

Example:

```
errorMessage={
 formikProps.getFieldMeta("username").value
  ? formikProps.errors.username
   : ""
}
```

**Comment by juuyan** [comment]
Any update or plan on this? Running into the same issue.

I think the current behavior is quite bad in UX. Would like to see the request behavior work as default, or at least give an option in <Formik/>

**Comment by cigzigwon** [comment]
I believe there is no solution for this. Only workarounds. I would run the latest version and disable validation so that fields only validate on submit. But I agree. Validations should probably default to validating only upon blur for that field. I think that validation runs on the entire schema for cases that involve cross field validation. Either way Formik will do the job. You just have to be dillegent in your own effort. I can help if you get stuck.

**Comment by cigzigwon** [comment]
What is the specific behavior you are looking for?

**Comment by cigzigwon** [comment]
@clu7ch3r Except I think you want to return undefined instead of an empty string because I beleive your field will be in error state with a blank message.

**Comment by juuyan** [comment]
@cigzigwon I think even for cross field validation, the current behavior is not the best way from the ux perspective. I'd expect something maybe like validate group.

I am simply building the behavior that validating only occurs upon blur for that field.

The problems I am running into is if there was anyway to build a workarounds with `<Field/>` and `<ErrorMessage/>`, in other words I would not have to pass all the event handlers manually...

And another question would be how to use Yup for field level validation...

Please give me some clue if you can.

**Comment by clu7ch3r** [comment]
@cigzigwon Mines work just fine. It only displays errors when there is one and the empty string is the placeholder space for when there is one.

**Comment by clu7ch3r** [comment]
@juuyan Have you tried my solution? So far it works on `TextInput` and RNE `Input` component.

**Comment by juuyan** [comment]
@clu7ch3r Hey I see your idea and will try it. But in your example you are not testing the change of the field's value, but its existence, aren't you?

**Comment by clu7ch3r** [comment]
@juuyan I was checking the change in the field's value. `formikProps.getFieldMeta("username").value` checks for change in username field, not if it exist. It would probably throw an error or do nothing if I passed in a field name that doesn't exist.

**Comment by juuyan** [comment]
@clu7ch3r 
Hmm but I have logged and `getFieldMeta(field-name).value` is the current value of field instead of the state of change. So I tried to store the value for comparing, yet have observed some side cases caused by repetitive state updating... Did I miss something?

**Comment by juuyan** [comment]
Share my workarounds here in case it could be helpful to someone.

I use a ref to store the state of field's blur, with which I determine if the error message should be visible or not.
```
const blur = useRef(false)

let errorMsg
if (blur.current) {
   errorMsg = errors[name] && errors[name]
}
```

And set the `blur` in `onBlur` and `onFocus`
```
<Field
   onBlur={e => {
      blur.current = true
      handleBlur(e)
   }}
   onFocus={() => (blur.current = false)}
/>
```

The benefit is you can still control the validate timing and schema with `<Formik/>` and `Yup`, but the validating will still execute on all the fields.

If you are looking for true single field validation, check this one:
https://github.com/jaredpalmer/formik/issues/512#issuecomment-473040060

**Comment by clu7ch3r** [comment]
@juuyan The value updates as you type so it would also be a state of change. I checked if the current value of the field has something in it and if it does, display the error for the specified field. With that being said, the error would only show up when typing in the specific field, not on all of them. Some previous comments above had already shown how to do it using Formik's `Field` component but my solution is one line of code and it's for those using the built-in `TextInput` component or third party ones like React Native Elements `Input` component.

**Comment by terenceodonoghue** [comment]
Is there a solution for this, or a plan to prevent blurring a single field from re-validating the entire form? I'm otherwise happy with Formik but this specific issue is infuriating.

**Comment by cigzigwon** [comment]
@terenceodonoghue Instead of getting so upset you might want to step back for a minute and re-consider your implementation. I may be able to help with that. How long have you been using Formik? I thought the same at first but now realize that @jaredpalmer has given us everything we need here to solve this. You should start by disabling validation altogether and leveraged the Formik touched prop to achieve what you are looking for. You can also just nix out Yup because we can use es6 to deal with validation of more complex data structuring that one may require. I'm not sure how complex you really need to be in working with your form. Is it a wizard? How many fields? I think the alternatives to Formik have other downfalls and that is why this project exists.

**Comment by terenceodonoghue** [comment]
My apologies @cigzigwon, I didn't mean to give the impression that I was upset.

My use case was validating the field value both synchronously (eg. 'value can't be empty') and asynchronously (validating value using a GraphQL hook)/

If the async validation failed it would assign the error to `asyncErrorStatus` which could then be read by the field's validate function, but then blurring an unrelated field would re-trigger validation on the original field and flag it as okay because the value was no longer empty.

I wound up solving my problem with `useEffect` like so:

```javascript
  useEffect(() => {
    validateForm();
  }, [asyncErrorStatus, validateForm]);
```

Basically, form validation behaviour proceeds as normal, but if the fetch error status changes, it triggers another validation and updates/clears field errors where appropriate.

Seems to work, but open to suggestions.

**Comment by vsubbotskyy** [comment]
To put my two cents in, I managed to validate fields separately by ditching `validationSchema` and using Formik's Field component and it's validate prop. And it works, but when I do:
```
await form.setFieldTouched(field.name, true, false);
await form.setFieldValue(field.name, value, false);
```
on input's _onChange_ for instance, I get very annoying lag between keypress and letter appearing in the input. Not awaiting is not an option because I'd be validating against old values.

**Comment by cigzigwon** [comment]
@vsubbotskyy I also chose not to rely on Yup because it has pros/cons depending on the Formik implementation. The lag is a side-effect of your synchronicity. Since Formik relies on ephemeral state you are blocking IO by overriding onChange directly. What problem are you trying to solve w/onChange in this way? Both of those things happen by default on Formik's onChange handler.

**Comment by cigzigwon** [comment]
@terenceodonoghue Ahh. Your challenge is dealing with race conditions from back end validation onBlur? That adds another level of difficulty, understood. I think whatever works to get you unblocked but in retrospect. Nix out ur complexity and take note that when passing a validator to the withFormik hook that you get all of the Formik props. So you can implement a filter to validate only touched fields instead of everything. I would use an object map reducer function pipeline to achieve this so you can iterate through touched fields and reduce the errors into the object Formik wants.

**Comment by cigzigwon** [comment]
```
const errors = Object.entries(values).filter(([key, value]) => (
	  	// filter out untouched and whatever you want to skip
	  ))
		.map(([key, value]) => {
	  	 // run validation and push error object defs into a stack
		})
		.reduce((acc, error) => {
	  	// reduce the array stack into the error obect formik wants
	  }, {})
return errors
```

**Comment by ivanfuzuli** [comment]
I use Field component with validation prop. but it also re-run all validations. Is there any work around? I make api calls and I don't wanna send request when other fields change.

I read a suggestion by @jaredpalmer that Field Level Validation will run only field change but it isn't solution. It re-run everytime. Do you have any example?

**Comment by adtm** [comment]
I today tackled also this `issue` and couldn't find a proper solution, but as current workaround came up just setting the fields to be touched on change. And showing the error only when they are touched. This prevents from checking all fields in one step and validates all of them on the submit. (It can be made to a simple wrapper even, but this gets the idea)

```js
<TextField
    placeholder={i18n.t("FIRST_NAME")}
    value={values.firstName}
    error={showErrorIfTouched(touched.firstName, errors.firstName)}
    onChangeText={handleChange("firstName")}
    onChange={() => setFieldTouched("firstName", true)}
    containerStyle={styles.input}
/>
```

**Comment by mateuspv** [comment]
Hey everyone, I'm running in the same issue:

- `handleChange` `handleBlur` should validate individual inputs.
- `onSubmit` should validate the entire schema.

`onSubmit`:
Should continue work as it is, validating all your `values` against `validationSchema`.

`handleChange` `handleBlur`:
Additionally, Inputs that have trigger `handleChange` and `handleBlur` will have their `values` tracked in another state.
`handleChange` `handleBlur` Should trigger validation only on those specific fields.

**Comment by cigzigwon** [comment]
@mateuspv Try without Yup. Use field level-only or write something functional and pass that as your validator. You can use map/reduce functions which are more efficient than Yup. Also account for blocking IO with async calls if you need back-end validations.

**Comment by mateuspv** [comment]
@cigzigwon Sorry I don't think it's about yup, is about be able of differentiate input validation from the entire form validation.

Even if i write my own validate function, is it running because it was trigger by `handleChange` or `handleBlur` ? Or it's because the user submitted the form(`onSubmit`)
```
validate: values => {
    const errors = {};
   
   // Did the user touched this? Is this a onSubmit or handleChange event ?
    if (!values.name) { 
      errors.name = 'Required';
    }

    // Same here
    if (!values.password) { 
      errors.password = 'Required';
    }

    return errors;
  }
```

Let me know if that makes sense for you,
appreciate your reply

**Comment by cigzigwon** [comment]
@mateuspv  Its not Yup. It still fires. But its okay to fire because its up to you to only display the errors when the field has been touched. Also theres nothing stopping you from checking touched internally. You can pass more than just values to validate. You can also use an external state to check via useState. Ultimately whats passed back in the errors strict is what controls display. Checks can go fast or slow depending on ur form field size. If the size is large then you should probably be building a wizard which will require a bit more dilligence to skip checks.

**Comment by cigzigwon** [comment]
@mateuspv To be specific,handleBlur behavior triggers setTouched on the field which then triggers the validation method OR Yup. handleSubmit would also do this validation. It all starts here: [https://jaredpalmer.com/formik/docs/api/formik#validateonblur-boolean](url)

Under the the process is as simple as outlined. Also in our forms we don't know if the user changed previous fields so I think that's why we just need to re-run validations. For your case you might want to look at withFormik so you can extend some of the behaviors outlined int the aforemention workflow. [https://jaredpalmer.com/formik/docs/api/withFormik](url)

**Comment by mateuspv** [comment]
@cigzigwon gotcha, I'm already using `withFormik `.
I found my issue: I was manually calling `validateForm` instead of `handleSubmit`.
This is why it was triggering all the validations.

Thanks for your help here! Apprecite.

**Comment by htkoca** [comment]
I've come to a similar solution to @adtm, while using material-ui and that is:
```
<TextField
  id="input-order-number"
  label="Order Number"
  type="orderNumber"
  name="orderNumber"
  onChange={handleChange}
  onBlur={handleBlur}
  value={values.orderNumber}
  error={Boolean(touched.orderNumber && errors.orderNumber)}
  helperText={
    (touched.orderNumber && errors.orderNumber) ||
    'Find it on your invoice'
  }
  fullWidth
/>
```

**Comment by alonspr** [comment]
This works for me (using validations + custom fields):
```
<Formik
    validate={validateInputs} // normal for loop which return an error object
    validateOnChange={false} // turn off fields validation
    validateOnBlur={false} // turn off fields validation
    ...
/>
```

I build my own fields, so for every field's onChange and onBlur I add this logic:
(validateInputs is the main validate function - you can pass it with context)
```
// saving new modified value
setFieldValue(name, e.target.value);

// validates local field only
// validateInputs - main validate func
const error = validateInputs({ [name]: e.target.value });
setFieldError(name, error[name]);
```
(for radio/checkbox you'll need to change props)

This way every change and blur validates the current field and not the entire form, this also keep the default behavior for submit / reset normally.

**Comment by alonspr** [comment]
This works for me (using validations + custom fields):
```
<Formik
    validate={validateInputs} // normal for loop which return an error object
    validateOnChange={false} // turn off fields validation
    validateOnBlur={false} // turn off fields validation
    ...
/>
```

I build my own fields, so for every field's onChange and onBlur I add this logic:
(validateInputs is the main validate function - you can pass it with context to components)
```
const validateSingleField = (name, value) {
    const error = validateInputs({ [name]: value });
    setFieldError(name, error[name]); // set / clear error message
};

const handleChange = e => {
    setFieldValue(name, e.target.value);

    // validate local field only
    validateSingleField(name, e.target.value);
};

const handleBlur = e => {
    setFieldTouched(name, true);

    // validate local field only
    validateSingleField(name, e.target.value);
};

<input onChange={handleChange} onBlur={handleBlur} {...} />

```
This way every change and blur validates the current field and not the entire form, this also keeps the default behavior for submit / reset normally.
You can of course set more conditions to avoid reassignment of same values

**Comment by cigzigwon** [comment]
@alonspr This is exactly what I was refferring to!

**Comment by 560560** [comment]
Hi everyone! This is my solution:

Validation starts work only after first press submit button. So user can safely filling Fields, and he get error messages only after trying of submitting . 

`import React, {useState} from 'react';
import {Button, Col, Container, Form, Row} from "react-bootstrap";
import {Formik} from "formik"
import * as Yup from 'yup'

const getSDataFromForm = (formDadta) => {
    console.log(formDadta)
}
const Contacts = (props) => {

    const [validationRequired, setValidationRequired] = useState(false)


    let initialValues = {
        firstName: "",
        secondName: "",
        email: "",
        message: "",
        checkbox: false
    }

    const validationSchema = Yup.object({
        firstName: Yup.string().required("This field is required").max(10),
        secondName: Yup.string().required("This field is required").max(12),
        email: Yup.string().required("This field is required").email("Please enter valid email"),
        message: Yup.string().max(12),
    })
return (
        <div>
            <Container style={{width: "500px"}}>
                <h1 className="text-center">Contact us</h1>
                <Formik initialValues={initialValues}
                        validationSchema={validationSchema}
                        onSubmit={getSDataFromForm}
                        validateOnChange={validationRequired}
                        validateOnBlur={validationRequired}
                         >
                    {({
                          handleSubmit,
                          handleChange,
                          isValid,
                          values,
                          touched,
                          errors
                      }) => (


                        <Form  noValidate onSubmit={handleSubmit} >

                            <Row className="mt-4 mb-3">

                                <Col>
                                    <Form.Label>First name</Form.Label>
                                    <Form.Control type="text"
                                                  name="firstName"
                                                  value={values.firstName}
                                                  placeholder="Enter your first name"
                                                  onChange={handleChange}
                                                  isValid={touched.firstName && !errors.firstName}
                                                  isInvalid={touched.firstName && !!errors.firstName}
                                    />

                                    <Form.Control.Feedback>Looks good!</Form.Control.Feedback>
                                    <Form.Control.Feedback type="invalid" className="justify-content-end d-flex">
                                        {errors.firstName}
                                    </Form.Control.Feedback>
                                </Col>

                                <Col>
                                    <Form.Label>Second name</Form.Label>
                                    <Form.Control type="text"
                                                  name="secondName"
                                                  value={values.secondName}
                                                  placeholder="Enter your second name"
                                                  onChange={handleChange}
                                                  isValid={touched.secondName && !errors.secondName}
                                                  isInvalid={touched.secondName && !!errors.secondName}
                                    />
                                    <Form.Control.Feedback>Looks good!</Form.Control.Feedback>
                                    <Form.Control.Feedback type="invalid" className="justify-content-end d-flex">
                                        {errors.secondName}
                                    </Form.Control.Feedback>
                                </Col>
                            </Row>
                            <Form.Group controlId="formBasicEmail">
                                <Form.Label>Email address</Form.Label>
                                <Form.Control type="email"
                                              name="email"
                                              value={values.email}
                                              placeholder="Enter email"
                                              onChange={handleChange}
                                              isInvalid={touched.email && !!errors.email}
                                              isValid={touched.email && !errors.email}
                                              />
                                <Row>
                                    <Col sm={7}><Form.Text muted>We'll never share your email with anyone
                                        else.</Form.Text></Col>
                                    <Col sm={5}>
                                        <Form.Control.Feedback type="invalid" className="justify-content-end d-flex ">
                                            {errors.email}
                                        </Form.Control.Feedback>
                                    </Col>
                                </Row>


                            </Form.Group>

                            <Form.Group controlId="formBasicMessage">
                                <Form.Label>Message</Form.Label>
                                <Form.Control as="textarea"
                                              name="message"
                                              rows="3"
                                              value={values.message}
                                              onChange={handleChange}
                                              isInvalid={touched.message && !!errors.message}
                                              />
                                <Form.Control.Feedback type="invalid" className="justify-content-end d-flex ">
                                    {errors.message}
                                </Form.Control.Feedback>
                            </Form.Group>
                            <Form.Group controlId="formBasicCheckbox">
                                <Form.Check type="checkbox"
                                            name="checkbox"
                                            label="Check me out"
                                            onChange={handleChange}

                                            />
                            </Form.Group>
                            <Button variant="primary" type="submit" onClick={()=>setValidationRequired(true)}>Submit</Button>
                        </Form>
                    )}
                </Formik>
            </Container>
        </div>
    );
}


export default Contacts;

`

**Comment by loburets** [comment]
I had a similar issue with validation of field value by API. It worked well, but user had to wait for API response even to see validation error for some other fields of the form.

To solve it I'm doing the async validation myself and put results to formik using setFieldError. But it also has few pitfalls, so maybe my example will be helpful for someone:

```
// component with the form:
const FormikExamplePage = (props) => {

  // use the apiError from component state as we need to have it stored outside of the formik
  // to not be reset after each validate cycle, but be reset only when we want
  const [apiErrorFromComponentState, setApiErrorToComponentState] = useState(null)

  return (
    <Formik
      validationSchema={yupSchema}
      validate={(values) => {
        const errors = {}

        // to not let formik reset it on validation
        if (apiErrorFromComponentState) {
          errors.firstNameApi = apiErrorFromComponentState
        }

        validateBirthdateBySeparateInputsAsTheSingleField(values, errors)

        return errors
      }}
    >
      {({
        errors,
        touched,
        isSubmitting,
        values,
        setFieldError,
        handleSubmit,
        setSubmitting,
        initialValues,
        setFieldTouched,
      }) => (
        <Form>
          Simple input + yup validation + api validation:<br/>
          <div>
            {/*
              * You can set the validate={validateFirstNameUsingApi} as props for formik.Field
              * It is the simplest solution which just works out of the box
              * But you maybe don't want to do it as it leads to delays during other validations
              * For example user edits the email field
              * Then the user see the error only when the api validation is done, despite yup rule doesn't need the api response to show the error
              * Even if the api validates the firstName, not the email, user still need to wait to see the email error
              * No solution currently is supported out of the box, see the https://github.com/formium/formik/issues/512
              * So, the component FirstNameFieldWithApiValidation is build to solve the issue
              * Also form submitting is overwritten to support it
              */}
            <FirstNameFieldWithApiValidation name="firstName" type="text" setApiError={setApiErrorToComponentState}/>
          </div>
          {/* Whatever logic of showing can be defined here, it is just example which looks reasonable for me */}
          { errors.firstName && touched.firstName ?
            <div style={{color: 'red'}}>{errors.firstName}</div>
              : errors.firstNameApi && touched.firstName ?
                  <div style={{color: 'red'}}>{errors.firstNameApi}</div>
                  : null
          }
            <Button
              type="submit"
              fullWidth
              color="primary"
              variant="raised"
              onClick={async e => {
                // overwritten formik submission just to add additional async validation before submitting
                e.preventDefault()

                if (isSubmitting) {
                  return
                }

                setSubmitting(true)
                // touch all fields to show the sync validation errors and don't force user to wait
                Object.keys(initialValues).forEach(key => setFieldTouched(key))

                const apiError = await validateFirstNameUsingApi(values.firstName)
                await setApiErrorToComponentState(apiError)
                setFieldError('firstNameApi', apiError)

                handleSubmit()
            }}>
              { isSubmitting ? 'Submitting...' : 'Submit' }
            </Button>
        </Form>
      )}
    </Formik>
  )
}


// example of field which is validated by api and yup both
const FirstNameFieldWithApiValidation = ({setApiError, ...props}) => {
  const {
    values: { firstName },
    errors: { firstName: firstNameError },
    setFieldError,
    setFieldValue,
    validateField,
  } = useFormikContext()

  const isMount = useIsMount()

  useEffect(() => {
    const doTheEffect = async () => {
      // update the field as touched but not on the first appearance
      // because we don't want to see the validation before user interacted with the field
      if (isMount) {
        return
      }

      // you can reset previous value if you want
      await setApiError(null)
      setFieldError('firstNameApi', null)
      // just to trigger rerender once again to reflect the error disappears till the new request will retur results
      setFieldValue('firstName', firstName)
      validateField('firstName')

      // some error is already here, so no need to validate on api
      // or maybe you want to do the api call and combine the errors, it's up to you
      if (firstNameError) {
        return
      }

      // you probably need some debounce mechanism here to not call api on each key down
      const apiError = await validateFirstNameUsingApi(firstName)
      setApiError(apiError)

      // the field is named as "firstNameApi" to understand difference if the field have error from api or not
      // it can help us to understand do we want to skip the api validation or not
      // otherwise we can not skip it based on the errors.firstName as it can be api error there which will not appear for the next api call
      setFieldError('firstNameApi', apiError)
    }
    doTheEffect()
  // set the value it only if something were changed, not the each render
  // it is also required to check if the firstNameError value was changed if you have the "if (firstNameError) { return }"
  // it works this way because value can be changes, but the error still not, so the validation would be skipped
  }, [firstName, firstNameError]);

  return (
    <Field {...props} />
  )
}
```

**Comment by JulkaIII** [comment]
@thenameisflic You solution is the most convenient and reusable!

**Comment by caduceusGithub** [comment]
The issue here is in our algorithm and our understand of how validation schema works.
I noticed that onChange the whole JSON object is validated every and all functions are executed.
So I added a simple decisional code:

![carbon](https://user-images.githubusercontent.com/69352895/93601564-e7fe8980-f9b0-11ea-8eef-2a487646a7fe.png)

I got around it without changing the way I implicated Formik.
Here **the back-end call** only happens **when the input modified**.

**Comment by cigzigwon** [comment]
@caduceusGithub That's why I said Yup < Functional Pipeline (Custom)

**Comment by appsparkler** [comment]
I'm working with `useFormik` with `Yup validationSchema` and this what worked for me:

```javascript
// inside the custom-hook
const validationSchema = Yup.object({
  userId: Yup
   .string()
   .test(async() =>  checkDuplicates())
   .isRequired(),
  firstName: Yup.string().isRequired()
})
const formik = useFormik({
  initialValues: {
    userId: '',
     firstName: ''
  },
  validationSchema,
  validateOnChange: false,
  validateOnBlur: false
})

// validations with useEffect
React.useEffect(() => {
  validationSchema.validateAt("userId", {
     userId: formik.values.userId
   })
    .catch(() => {
     //handle error for user-id validation
    });
},[formik.values.userId])
React.useEffect(() => {
   validationSchema.validateAt("firstName", {
     userId: formik.values.firstName
   })
    .catch(() => {
     //handle error for firstName validation
    });
},[formik.values.firstName])
```

```jsx
// jsx snippet of the form
<>
  <input value={formik.values.firstName} 
    placeholder="First Name. ..." 
    onChange={formik.handleChange}
  />
  <input 
    value={formik.values.userId} 
    placeholder="User ID..." 
     onChange={formik.handleChange}
  />
</>
```

**Comment by JaredDahlke** [comment]
yeah i just noticed it was validating all fields every keystroke when i throttled my CPU down 4x and put some console logs in my validation function.

this is my validationSchema , the topics array has over 2000 objects in it, definitely going to have to find yet another work around to get formik workin on this:

you can see here why validating every field wont work for me. So if i enter in a 5 character profileName, my topicsHasResponse function runs over 1000 times :
```export const schemaValidation = Yup.object().shape({
	basicInfoIndustryVerticalId: Yup.number()
		.typeError('Required')
		.required('Required'),
	basicInfoProfileName: Yup.string()
		.min(2, 'Must be greater than 1 character')
		.max(50, 'Must be less than 50 characters')
		.required('Required'),
	basicInfoWebsiteUrl: Yup.string()
		.test(
			'urlTest',
			'Valid URL required (e.g. google.com)',
			(basicInfoWebsiteUrl) => {
				return urlRegex({ exact: true, strict: false }).test(
					basicInfoWebsiteUrl
				)
			}
		)
		.required('Required'),

	basicInfoTwitterProfile: Yup.string()
		.min(2, 'Must be greater than 1 character')
		.max(50, 'Must be less than 30 characters')
		.required('Required'),
	topCompetitors: Yup.array()
		.typeError('Wrong type')
		.min(1, 'At least one competitor is required'),
	topics: Yup.array()
		.typeError('Wrong type')
		.test('topicsTest', 'You must include at least one topic', (topics) => {
			return topicsHasResponse(topics)
		}),
	scenarios: Yup.array()
		.typeError('Wrong type')
		.test(
			'scenariosTest',
			'Please select a response for each scenario',
			(scenarios) => {
				return scenariosAllHaveAResponse(scenarios)
			}
		),
	categories: Yup.array()
		.typeError('Wrong type')
		.test(
			'categoriesTest',
			'Please take action on at least one category',
			(categories) => {
				return categoriesHasResponse(categories)
			}
		)
})

function categoriesHasResponse(categories) {
	console.log('running categoriesHasResponse')
	if (categories.length < 1) return false
	for (const category of categories) {
		if (category.contentCategoryResponseId !== 3) return true
	}
	return false
}

function topicsHasResponse(topics) {
	console.log('running topics has response')
	for (const topic of topics) {
		console.log('looping through topics')
		if (topic.topicResponseId == 1) return true
		if (topic.children && topic.children.length > 0) {
			const childHasResponse = topicsHasResponse(topic.children)
			if (childHasResponse) return childHasResponse
		}
	}
	return false
}

function scenariosAllHaveAResponse(scenarios) {
	console.log('running scenaios all have a response')
	if (scenarios.length < 1) return false
	for (const scenario of scenarios) {
		console.log('looping through scenarios')
		if (!scenario.scenarioResponseId || scenario.scenarioResponseId.length < 1)
			return false
	}
	return true
}```

**Comment by johnrom** [comment]
@JaredDahlke it's definitely understandable that this is a pretty severe limitation for some large datasets. the problem with solving it by validating only one field at a time, is that some fields might have dependencies causing the following issue:

```js
// passed to fieldB.validate prop
fieldBisValid = (fieldA * fieldB) < 100;

// the validation here is correct in both systems
setValues({
  fieldA: 20,
  fieldB: 4,
});

// the validation run here currently works,
// but would not if we only validated fieldA.
setFieldValue('fieldA', 40, true);
```

The problem is how do we provide a way for one field-level validation to trigger another in a way that makes sense. I think an API like this makes some sense, but requires just gallons of other work to be done before implementing and without proper care could lead to an infinite loop of validations if both fields include each other.

```
// includes values.fieldA, touched.fieldA, errors.fieldA, and triggers fieldA.validations when these validations are triggered
<Field name="fieldB" include={fields => { fieldA: fields.fieldA }} />
```

**Comment by itachiluan** [comment]
My particular need is just to make sure that the validator doesn't go to the API unnecessarily, so I've created a simple class that creates a state that caching the previous result, which, by it self, would also check email validity first (somehow through validationSchema, the validation of Yup.test() still fires up even the email value is not valid):
```
const emailSchema = Yup.string().email().required();

export class UniqueEmailTester {

    static state() {
        return {
            lastState: {
                email: '',
                result: false
            }
        }
    }

    static async validator(email) {

        if (this.lastState.email === email)
            return this.lastState.result;

        this.lastState.email = email;
        try {
            await emailSchema.validate(email);
        } catch {
            this.lastState.result = false;
            return false;
        }

        try {
            const response = await Axios.request({
                baseURL: 'http://some.domain/api',
                url: '/user/uniqueemail',
                method: 'post',
                data: { email }
            });

            this.lastState.result = response.data.isValid;

            return response.data.isValid;
        } catch {
            this.lastState.result = false;
            return false;
        }

    }
}

```

then in the component I could just use it like this:
```
let emailValidatorState = UniqueEmailTester.state();

const validationSchema = Yup.object().shape({
        email: Yup.string().required("Email address is required.")
            .email("Invalid email address.")
            .lowercase()
            .test('checkEmailUnique', 'This address has already been registered with us', 
                UniqueEmailTester.validator.bind(emailValidatorState)),
        ...some_other_fields
})

```

**Comment by jesseko** [comment]
Changing this as proposed would also help a ton with server errors.

Jared wrote: "If you use setErrors, your errors will be wiped out by Formik's next validate or validationSchema call which can be triggered by the user typing (a change event) or blurring an input (a blur event)" ([SO link](https://stackoverflow.com/questions/52986962/how-to-properly-use-formiks-seterror-method-react-library))

But if that validation was only per-field then only the active field's error would get cleared, the rest of the field-level server errors would persist.  Definite behavior improvement.

---

@johnrom would it be worth considering declaring the dependencies closer to the validation rather than directly on the Fields?

Not sure if Yup structures/bindings would allow for unpacking something like this, but just a thought:
```
validationSchema: Yup.object({
  isBig: boolean(),
  count: {  // optionally provide an object of this shape 
     dependencies: [ 'isBig' ],  // tell Formik to also validate this field when any of its dependencies change
     test: number().when('isBig', {   // the normal stuff you'd do with yup
        is: true,
        then: yup.number().min(5),
        otherwise: yup.number().min(0),
    }), ...
```

**Comment by johnrom** [comment]
@jesseko I don't know if Yup supports that, sorry (I don't use it).

The important determinations for making this change a reality are:

### An Opt-In method

We need to provide a way to opt in to this new functionality to not break it for people using dependent fields in the current environment. We could either 

- provide a prop like `<Formik isolatedFields={true} />`
- create a new IsolatedField (or something) which does the above to not cause breaking changes to FastField.

### Expressing dependent fields

We need a way to express field dependencies. Some options:

- Yup dependencies option above (not recommended as we shouldn't track Yup features in Formik)
- Field `include={}` api, I recommend although it would serve two purposes which might be confusing -- to add dependent field state to current field and also trigger validation when that state changes.
- Some other way

### Conclusion

When we determine paths for each of the above, we can implement something.

**Comment by jesseko** [comment]
FWIW I was suggesting that Yup structure with the idea that it's be extra syntax on the Formik side, and Formik would process it to create 1) an internal structure representing those dependencies and 2) the Yup schema exactly as it looks today to actually execute validation.  

Point taken about handling dependent fields as well.  Looking at the [dependent fields example](https://formik.org/docs/examples/dependent-fields-async-api-request), I see that `MyField` (`textC`) is getting values for fields it cares about (`textA, textB`) from `useFormikContext`.  

Is the concern that with the optimizations in Formik V3, changes to `textA` and `textB` won't cause `MyField` (`textC`) to re-render without something like your `include={}` idea above to tell Formik about those dependencies?

**Comment by johnrom** [comment]
In v3, useFormikContext doesn't contain state, so the equivalent would be `useFieldMeta()`, but you cannot randomly add hooks depending on props, so that needs to itself be a prop like `include` that can define additional state to include within a single subscription.

It's a little convoluted, but yes basically the same thing but very optimized.

**Comment by parker-upstart** [comment]
Has this issue still yet to be resolved? 100's of comments on this issue dating back over 2 years as well, I must be missing something?

**Comment by cigzigwon** [comment]
@parker-upstart I've been using Formik for a while... Thought this WAS an issue but it's not. You need to understand how to disable uatomatic bindings and then do your own implementation. Yup seems to stand in the way. Use my own validator and it's less code and 10x faster. This issue should be closed...

**Comment by parker-upstart** [comment]
@cigzigwon Yea, this is what I'm doing using a custom `validate` function - but with all the praise around how yup in tightly integrated I feel like this is still something which users would like to have. If that's not the direction the devs want to go that's fine, was mainly curious if I missed something in this thread.

**Comment by nycgavin** [comment]
I feel like if formik provided an implmentation for this, we don't have to roll our own implementation

**Comment by nycgavin** [comment]
I think currently formik is missing an api where the validation doesn't validate all fields.  If I can do something like this: helper.validate(fieldname) I can roll my own code, and still use setError and custom field validation method in formik, if I do this setValue(fieldName, value, true) every other field gets validated

**Comment by kafeelkhatri** [comment]
So onething what i found out is that using touched feature that will validate on handleSubmit and after that it will work perfectly by validating every field 
{({ handleChange, handleSubmit, errors,  values, isValid, **touched**,})

Give it a try maybe it will help

**Comment by ldwrs** [comment]
Ideally I’d have a yup schema and only the field that’s been changed gets validated… Maybe this isn’t the best place to ask, but could React Hook Forms do this?

**Comment by npearson72** [comment]
@ldwrs I'm currently running into this issue with react hook form. Haven't found a solution yet.

**Comment by patrickyan** [comment]
@ldwrs Can confirm that React Hook Form runs the test on every keystroke as well

**Comment by wallace-sf** [comment]
I'm currently creating a form step and before to switch page i want to validate page fields. So, i want validate one field at time... Formik does not provide a simple way to do that.

**Comment by cigzigwon** [comment]
When dealing with Formik customization that's specialized we don't use Yup and we disable validateOnBlour valdateOnChange

And use a functional pipeline w/reducer:

```
export const validate = (values, tabs, index, uppy) => {
  const errors = Object.entries(values).filter(([key, value]) => (
  	key.match(tabs[index]) !== null // match tabname
  ))
	.map(([key, value]) => {
  	let error = {}
  	const elem = document.getElementById(key)export const validate = (values, tabs, index, uppy) => {
  const errors = Object.entries(values).filter(([key, value]) => (
  	key.match(tabs[index]) !== null // match tabname
  ))
	.map(([key, value]) => {
  	let error = {}
  	const elem = document.getElementById(key)
  	const select = elem ? M.FormSelect.getInstance(elem) : null
  	const validator = validators[key] ? validators[key] : null
  	// run validation per field key
  	if (validator) {
  		error[key] = validator(value)
  	}
  	// change class on M.Select (menu)
		if (error[key] && select) {
			select.wrapper.className = select.wrapper.className + ' invalid'
		} else if (select) {
			select.wrapper.className = 'select-wrapper'
		}
		return error
	})
	.reduce((acc, error) => {
  	const key = Object.keys(error)[0]
  	if (!error[key]) {
  		return acc
  	}
  	acc[key] = error[key] 
  	return acc
  }, {})

	if (Object.keys(errors).length) {
		console.log('Formik errors', errors)
		scroll_reset()
	}

	if (index + 1 >= tabs.length) {
		const valid = valid_files(uppy)
		if (valid !== undefined) {
			return {...errors, ...{ docs_checklist: valid }}
		}
	}
  
  return errors
}
  	const select = elem ? M.FormSelect.getInstance(elem) : null
  	const validator = validators[key] ? validators[key] : null
  	// run validation per field key
  	if (validator) {
  		error[key] = validator(value)
  	}
  	// change class on M.Select (menu)
		if (error[key] && select) {
			select.wrapper.className = select.wrapper.className + ' invalid'
		} else if (select) {
			select.wrapper.className = 'select-wrapper'
		}
		return error
	})
	.reduce((acc, error) => {
  	const key = Object.keys(error)[0]
  	if (!error[key]) {
  		return acc
  	}
  	acc[key] = error[key] 
  	return acc
  }, {})

	if (Object.keys(errors).length) {
		console.log('Formik errors', errors)
		scroll_reset()
	}

	if (index + 1 >= tabs.length) {
		const valid = valid_files(uppy)
		if (valid !== undefined) {
			return {...errors, ...{ docs_checklist: valid }}
		}
	}
  
  return errors
}
```

**Comment by cigzigwon** [comment]
@wallace-sf So much faster this way. You just need to call the validators you define per trigger event that you override and then you can replace Yup w/your own functionaly reducer. My form has six tabs and varying workflows. On submit fires on every tab change foraward or backawards. It does everything and also has an Uppy hook in the form. Sometimes you can cook up your own function  that's thrity lines of code that does exactly what you need to do.

**Comment by cigzigwon** [comment]
validators.js (my code is very modular)

```
export const is_one = value => (
	!value ? "Choose at least one" : undefined
)
export const is_required = value => (
	!value ? "Field is required" : undefined
)
export const multi_one = values => (
	!values.filter(value => (
		value !== ""
	)).length ? "Choose at least one" : undefined
)
export const min_length = (value, length) => (
	value.length < length ? `Minimum of ${length} characters in length` : undefined
)
```

**Comment by ldwrs** [comment]
@cigzigwon I don't think this is possible when your validation rules depend on other field values?

**Comment by cigzigwon** [comment]
@ldwrs No you are absolutely incorrect. Use ingenuity.

**Comment by ldwrs** [comment]
You'll be more convincing with a working example...

**Comment by ldwrs** [comment]
I think you're missing the point. You can validate the whole form on submit of course. But what we need is validating one field only onchange (immediately) with validation rules that depends on other field values.

**Comment by cigzigwon** [comment]
@ldwrs My working example is too complex to share. My point is that if you have skills put them to use. If not then, I guess you are stuck aren't you?

**Comment by ldwrs** [comment]
Ok, I see you can not give an example of what is asked. I think anyone can draw the correct conclusions from this simple fact.

**Comment by Zach-Jaensch** [comment]
> @ldwrs No you are absolutely incorrect. Use ingenuity.

Your response is incredibly rude, and if you are not able to provide a simplified example, I could argue that you don't have the skills needed.



> @cigzigwon I don't think this is possible when your validation rules depend on other field values?

We don't use yup in our project (sadly not my choice) but we have been able to validate fields on an individual level, even when they are coupled to other fields. I can imagine a similar concept might be capable with yup, but you may have to change how you implement it. 

(Note: parts unrelated to validation are removed)
We created a "wrapping" component so we can compose any input type with it
```typescript
function FormikField(props) {
  const Component: any = props.component || BasicInput;

  const { values } = useFormikContext();

  function handleValidation(value) {
    if (validation) {
      return await validation(value, values) || undefined;
    }

    return undefined;
  }

  const [field] = useField({
    name: props.name,
    validate: handleValidation
  });

  return <Component {...field} />
}
```

**Comment by cigzigwon** [comment]
I'm simply not doing your thinking and work. If you guys knew how to combine functional pipelines with a decorator pattern. This wouldn't be an issue for you today. I e migrated everything into Phoenix LiveView by now so Formik is now obsolescent in our apps since December 25th. There's nothing you could say to make me think less of myself. I'm just amused.

**Comment by Zach-Jaensch** [comment]
This is a community page centred around formik and the issue clearly many people have. Please stick to the subject of supporting the community and not using it as a forum to demonstrate your ego. To help build a community (and a team) of great engineers, the knowledgable people should always be willing to help, not hinder and put down as you seam to be attempting.

**Comment by cigzigwon** [comment]
My original counter argument is that it's not really an issue if you just start writing code. At some point you stop copy/pasting the work of a great engineer.

**Comment by cigzigwon** [comment]
My best advice for this lib as working functionally is that the common structure of errors be changed to an array of objects with support for a functional return. And that Yup be not used. And the moving everything so that an array of objects that have functional returns as validators. The fields that depend on fields are one of but can b dealt with in the functional return pattern as they can reach into another validator. Forms are lists of lists. So array of object patterns would have been ideal. The key to get what you need from Formik is flexibility and congruency between lists of lists. A common structure in a functional language which JavaScript is. I could probably just take this summation to make a simple port for LiveView. Kinda already has this out of the box though. I'm unsubscribing as going any deeper is a waste of my time

**Comment by meysamzi** [comment]
How can i have a real-time validation with Formik?
i mean when a user types in an input field not when he/she clicks on Submit button or when onBlur happens!

**Comment by kjschulz** [comment]
Better late than never 😂??? Thanks to inspiration from @pedroabreu and @cigzigwon, I was able to get this working with a normal Formik/React Bootstrap (BS v5) install with a `validationSchema` that utilizes `yup` without too much customization...in case this helps anyone else.  This will allow each field to be validated, one at a time, as changes are made and also still performs the default whole form validation on each form submit:

```javascript
<Formik
  initialValues={myInitialValues}
  validationSchema={myYupSchema}
  validateOnChange={false} // Disable validation every field change
  validateOnBlur={false} // Disable validation every field blur
>
  {({
    handleSubmit,
    handleChange,
    handleBlur,
    validateField,
    values,
    touched,
    isValid,
    errors
  }) => {
    // Avoid a race condition to allow each field to be validated on change
    const handleInputChange = async (e, fieldName) => {
      await handleChange(e);
      validateField(fieldName);
    };

    return (
      <Form noValidate onSubmit={handleSubmit}>
        <Row className="mb-3">
          <Form.Group as={Col} md="6" controlId="validateFirstName">
            <Form.Label>First Name</Form.Label>
            <Form.Control
              type="text"
              name="firstName"
              value={values.firstName}
              onChange={(e) => handleInputChange(e, 'firstName')}
              isValid={touched.firstName && !errors.firstName}
              isInvalid={!!errors.firstName}
            />
              ...
      </Form>
    )
  }}
</Formik>
```

**Comment by ml054** [comment]
Not sure if you know but formik is dead for years, so maybe it is time for change? ;-) New season new form library. Welcome in JS world. :)

**Comment by kjschulz** [comment]
Hmm, ok thanks @ml054.  I noticed that the project is not currently being maintained but I guess I didn't realize it was **dead**.  It's still listed in the current documentation as the main suggested option in React-Bootstrap's latest version and on many React form validation library round ups form just last year.

Out of curiosity though, what would you recommend?

**Comment by ml054** [comment]
I'd say https://npmtrends.com/ to avoid doing advertisement. ;-)

**Comment by wallace-sf** [comment]
> @wallace-sf So much faster this way. You just need to call the validators you define per trigger event that you override and then you can replace Yup w/your own functionaly reducer. My form has six tabs and varying workflows. On submit fires on every tab change foraward or backawards. It does everything and also has an Uppy hook in the form. Sometimes you can cook up your own function that's thrity lines of code that does exactly what you need to do.

That's right. I made a first version, as you said. Later, for other page steps, I decided to use a `Formik` component for each page. No parent `Formik `component for pages. It was easier that way.

**Comment by wallace-sf** [comment]
> @wallace-sf So much faster this way. You just need to call the validators you define per trigger event that you override and then you can replace Yup w/your own functionaly reducer. My form has six tabs and varying workflows. On submit fires on every tab change foraward or backawards. It does everything and also has an Uppy hook in the form. Sometimes you can cook up your own function that's thrity lines of code that does exactly what you need to do.

@cigzigwon That's right. I made a first version, as you said. Later, for other page steps, I decided to use a Formik component for each page. No parent Formik component for pages. It was easier that way.

**Comment by cigzigwon** [comment]
@wallace-sf Yes, all it really is a primary validation function (decorator) for onSubmit actions and then for "one filed validation" simply apply it to that field or inject it into custom Component. We should alsways use hooks in this lib and avoid Yup. When the use cases evolve beyond smaller digital ,marketing forms. Yup would also work fine in most back-end data entry forms where the entmry is more straightforward and less dependant on other field requirements. I do appreciate you for leaving a comment today. At one point some wanted more examples but I do not have time to provide them but this is my last comment in this thread. Thanks!

**Comment by cigzigwon** [comment]
@kjschulz I agree that's the defacto way to overide the underlying Promises with async/await in the handler but some might be missing that point here. I spent a while on it myself.

**Comment by Ismail0v12** [comment]
Hello everone, here i am coming with solution, in formik has function validateField(value: string), so if you pass this function to props onBlur or onChange, it is validating only passed field to validateField, like this:
`      onBlur={() => {
          registerForm.validateField('confirm_password');
       }}
`
if you have better solution with validateField function just let us know!
Happy coding :)

**Comment by NicoConstantin** [comment]
> Better late than never 😂??? Thanks to inspiration from @pedroabreu and @cigzigwon, I was able to get this working with a normal Formik/React Bootstrap (BS v5) install with a `validationSchema` that utilizes `yup` without too much customization...in case this helps anyone else. This will allow each field to be validated, one at a time, as changes are made and also still performs the default whole form validation on each form submit:
> 
> ```js
> <Formik
>   initialValues={myInitialValues}
>   validationSchema={myYupSchema}
>   validateOnChange={false} // Disable validation every field change
>   validateOnBlur={false} // Disable validation every field blur
> >
>   {({
>     handleSubmit,
>     handleChange,
>     handleBlur,
>     validateField,
>     values,
>     touched,
>     isValid,
>     errors
>   }) => {
>     // Avoid a race condition to allow each field to be validated on change
>     const handleInputChange = async (e, fieldName) => {
>       await handleChange(e);
>       validateField(fieldName);
>     };
> 
>     return (
>       <Form noValidate onSubmit={handleSubmit}>
>         <Row className="mb-3">
>           <Form.Group as={Col} md="6" controlId="validateFirstName">
>             <Form.Label>First Name</Form.Label>
>             <Form.Control
>               type="text"
>               name="firstName"
>               value={values.firstName}
>               onChange={(e) => handleInputChange(e, 'firstName')}
>               isValid={touched.firstName && !errors.firstName}
>               isInvalid={!!errors.firstName}
>             />
>               ...
>       </Form>
>     )
>   }}
> </Formik>
> ```

This works also with onBlur event, if you have your form modularized you can access to handleChange or handleBlur and also to validateField from form parameter:
Example:
<Field
        name={data.name}
        type={data.type}
      >
        {({ field, form }) => (
            <input
              className={inputClass(errors[data.name])}
              type={data.type}
              {...field}
              onBlur={async e => {
                await form.handleBlur(e)
                form.validateField(data.name)
              }}
            />
        )}
      </Field>

**Comment by getsnoopy** [comment]
I don't understand why this behaviour isn't considered a bug; I just had to go down a debugging rabbit hole and try all sorts of things to make sure I wasn't calling the same validator function multiple times or the like to see why my validator functions for other fields were running when an unrelated field changed. The [validation documentation](https://formik.org/docs/guides/validation#form-level-validation) clearly says:

> Form-level validation is useful because you have complete access to all of your form's values and props whenever the function runs, so you can **validate dependent fields at the same time**.

This implies that non-form-level validation (i.e., *field*-level validation) runs at the individual field level, as the name implies, and isn't intended for dependent field validation. So why is that being used as a reason for this behaviour? It seems like dependent fields are in the minority of cases rather than the majority, yet this silly behaviour is leading to the majority having to implement all sorts of gnarly workarounds.

Having to disable default validation/event handling functionality, reroute those calls to a custom event handler function that then manually calls the `handleChange` function and then calls the individual field validator function entirely defeats the purpose of the purported "composability" and avoidance of "prop-drilling" that Formik apparently provides according to its [headline tutorial page](https://formik.org/docs/tutorial#leveraging-react-context).

The solution seems clear to me: just disable this behaviour. If people want dependent field validation, they can either do that at the form validation level, or we can implement some options called `validateAllFieldsOnChange`, `validateAllFieldsOnBlur`, etc. or something to that effect that enables the current behaviour for them.

**Comment by amidulanjana** [comment]
I had the same issue; it runs the test method on every keystroke. So, I went with useRef. It prevents the unnecessary running of the test method when other fields change. 

```
.test('domain-server-validation', 'This domain is already in use', async domain => {
    if (domain && domainFieldValidationRef.current !== domain) {
      domainFieldValidationRef.current = domain;
      .....
      return false
    }
    return true
 })
```

**Comment by dribble-njr** [comment]
I find a solution in [formik demo](https://formik.org/docs/examples/with-material-ui). You can see the effect by setting the initial value of the form to empty.

**Comment by 7vazier** [comment]
You can just simply check if an input is given as well as if error is generated or not 

like this

 {values.First_field && errors.First_field ? <p>errors.First_field</p>: ""}
